File: comfyui-edition/ui/src/services/expressions/motionExpressions.ts
/**
 * Motion Expressions Module
 *
 * Expression functions for momentum-based animations:
 * - inertia: Momentum after keyframes with decay
 * - bounce: Bouncing settle physics
 * - elastic: Spring-like elastic motion
 */

import type { Keyframe } from "@/types/project";
import { isFiniteNumber } from "@/utils/typeGuards";

/**
 * Minimal expression context for motion expressions
 * (Avoids circular dependency with main expressions.ts)
 */
export interface MotionExpressionContext {
  time: number;
  fps: number;
  keyframes: Keyframe<number | number[]>[];
  value: number | number[];
  velocity: number | number[];
}

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
//                                                                   // helpers
// ════════════════════════════════════════════════════════════════════════════

/**
 * Convert value to array format for consistent processing.
 * Handles both array [x, y, z] and object {x, y, z} formats.
 * Prevents crashes with object-style vectors by normalizing to arrays.
 */
function toArray(value: number | number[] | { x: number; y: number; z?: number }): number[] {
  if (typeof value === 'number') {
    return [value];
  }
  if (Array.isArray(value)) {
    return value;
  }
  // Object format {x, y} or {x, y, z}
  if (typeof value === 'object' && value !== null && 'x' in value && 'y' in value) {
    const arr = [value.x, value.y];
    if ('z' in value && value.z !== undefined) {
      arr.push(value.z);
    }
    return arr;
  }
  // Fallback: return as single-element array or empty
  return [0];
}

/**
 * Convert array back to original format based on original value structure
 */
function fromArray(
  arr: number[],
  originalValue: number | number[] | { x: number; y: number; z?: number }
): number | number[] | { x: number; y: number; z?: number } {
  if (typeof originalValue === 'number') {
    // Type proof: arr[0] ∈ number | undefined → number
    return isFiniteNumber(arr[0]) ? arr[0] : 0;
  }
  if (Array.isArray(originalValue)) {
    return arr;
  }
  // Object format
  if (typeof originalValue === 'object' && originalValue !== null && 'x' in originalValue) {
    // Type proof: arr[0] ∈ number | undefined → number
    const x = isFiniteNumber(arr[0]) ? arr[0] : 0;
    // Type proof: arr[1] ∈ number | undefined → number
    const y = isFiniteNumber(arr[1]) ? arr[1] : 0;
    const result: { x: number; y: number; z?: number } = {
      x,
      y,
    };
    if ('z' in originalValue && originalValue.z !== undefined) {
      // Type proof: arr[2] ∈ number | undefined → number
      result.z = isFiniteNumber(arr[2]) ? arr[2] : 0;
    }
    return result;
  }
  return arr;
}

// ════════════════════════════════════════════════════════════════════════════
//                                                     // motion // expressions
// ════════════════════════════════════════════════════════════════════════════

/**
 * Inertia/Overshoot expression
 * Creates momentum-based animation after keyframes
 */
export function inertia(
  ctx: MotionExpressionContext,
  amplitude: number = 0.1,
  frequency: number = 2.0,
  decay: number = 2.0,
): number | number[] {
  const { time, keyframes, value, velocity } = ctx;

  // Validate parameters
  const safeAmplitude = Number.isFinite(amplitude) ? amplitude : 0.1;
  const safeFrequency = Number.isFinite(frequency) ? frequency : 2.0;
  const safeDecay = Number.isFinite(decay) ? Math.max(0.001, decay) : 2.0;

  if (keyframes.length === 0) return value;

  // Find nearest keyframe before current time
  // Type proof: fps ∈ number | undefined → number
  const fps = isFiniteNumber(ctx.fps) && ctx.fps > 0
    ? ctx.fps
    : 16;
  const currentFrame = time * fps;

  let nearestKey: Keyframe<number | number[]> | null = null;
  for (let i = keyframes.length - 1; i >= 0; i--) {
    if (keyframes[i].frame <= currentFrame) {
      nearestKey = keyframes[i];
      break;
    }
  }

  if (!nearestKey) return value;

  const keyTime = nearestKey.frame / fps;
  const t = time - keyTime;

  if (t <= 0) return value;

  // Convert to arrays for consistent processing
  const valueArr = toArray(value);
  const velocityArr = toArray(velocity);

  // Apply oscillation to each component
  const resultArr = valueArr.map((v, i) => {
    // Type proof: componentVel ∈ number | undefined → number
    const componentVel = isFiniteNumber(velocityArr[i]) ? velocityArr[i] : 0;
    const oscillation =
      (componentVel * safeAmplitude * Math.sin(safeFrequency * t * 2 * Math.PI)) /
      Math.exp(safeDecay * t);
    return v + oscillation;
  });

  // Convert back to original format
  return fromArray(resultArr, value) as number | number[];
}

/**
 * Bounce expression
 * Creates bouncing settle after keyframes
 */
export function bounce(
  ctx: MotionExpressionContext,
  elasticity: number = 0.7,
  gravity: number = 4000,
): number | number[] {
  const { time, keyframes, value } = ctx;

  // Validate parameters
  const safeElasticity = Number.isFinite(elasticity)
    ? Math.max(0, Math.min(1, elasticity))
    : 0.7;
  const safeGravity = Number.isFinite(gravity) && gravity > 0 ? gravity : 4000;

  if (keyframes.length === 0) return value;

  // Type proof: fps ∈ number | undefined → number
  const fps = isFiniteNumber(ctx.fps) && ctx.fps > 0
    ? ctx.fps
    : 16;
  const currentFrame = time * fps;

  // Find last keyframe
  let lastKey: Keyframe<number | number[]> | null = null;
  for (let i = keyframes.length - 1; i >= 0; i--) {
    if (keyframes[i].frame <= currentFrame) {
      lastKey = keyframes[i];
      break;
    }
  }

  if (!lastKey) return value;

  const keyTime = lastKey.frame / fps;
  const t = time - keyTime;

  if (t <= 0) return value;

  // Bounce physics
  let bounceTime = t;
  let bounceHeight = 1;
  let totalBounces = 0;
  const maxBounces = 10;

  // Calculate which bounce we're in
  while (bounceTime > 0 && totalBounces < maxBounces) {
    const bounceDuration = Math.sqrt((2 * bounceHeight) / safeGravity);
    if (bounceTime < bounceDuration * 2) {
      break;
    }
    bounceTime -= bounceDuration * 2;
    bounceHeight *= safeElasticity * safeElasticity;
    totalBounces++;
  }

  // Position within current bounce
  const bounceDuration = Math.sqrt((2 * bounceHeight) / safeGravity);
  const bounceT = bounceTime / (bounceDuration * 2);
  const bounceOffset = bounceHeight * 4 * bounceT * (1 - bounceT);

  // Convert to arrays for consistent processing
  const valueArr = toArray(value);
  const resultArr = valueArr.map((v) => v - bounceOffset * (1 - safeElasticity));
  
  return fromArray(resultArr, value) as number | number[];
}

/**
 * Elastic expression
 * Creates elastic spring-like motion
 */
export function elastic(
  ctx: MotionExpressionContext,
  amplitude: number = 1,
  period: number = 0.3,
): number | number[] {
  const { time, keyframes, value } = ctx;

  // Validate parameters
  const safeAmplitude = Number.isFinite(amplitude) ? amplitude : 1;
  const safePeriod = Number.isFinite(period) && period > 0 ? period : 0.3;

  if (keyframes.length === 0) return value;

  // Type proof: fps ∈ number | undefined → number
  const fps = isFiniteNumber(ctx.fps) && ctx.fps > 0
    ? ctx.fps
    : 16;
  const currentFrame = time * fps;

  let lastKey: Keyframe<number | number[]> | null = null;
  for (let i = keyframes.length - 1; i >= 0; i--) {
    if (keyframes[i].frame <= currentFrame) {
      lastKey = keyframes[i];
      break;
    }
  }

  if (!lastKey) return value;

  const keyTime = lastKey.frame / fps;
  const t = time - keyTime;

  if (t <= 0) return value;

  const s = safePeriod / 4;
  const decay = 2 ** (-10 * t);
  const oscillation = decay * Math.sin(((t - s) * (2 * Math.PI)) / safePeriod);

  // Convert to arrays for consistent processing
  const valueArr = toArray(value);
  const resultArr = valueArr.map((v) => v + safeAmplitude * oscillation);
  
  return fromArray(resultArr, value) as number | number[];
}

File: leancomfy/haskell/Lattice/Services/Expressions/MotionExpressions.hs
{-# LANGUAGE StrictData #-}

{-|
Module      : Lattice.Services.Expressions.MotionExpressions
Description : Motion physics functions
Copyright   : (c) Lattice, 2026
License     : MIT

Core physics calculations for momentum-based animations:
- inertiaOscillation: Damped sine wave oscillation
- bouncePhysics: Bouncing settle with elasticity
- elasticOscillation: Spring-like elastic motion

Source: ui/src/services/expressions/motionExpressions.ts
-}

module Lattice.Services.Expressions.MotionExpressions
  ( -- * Constants
    defaultAmplitude, defaultFrequency, defaultDecay
  , defaultElasticity, defaultGravity, defaultPeriod
  , maxBounces
    -- * Inertia
  , inertiaOscillation
  , inertiaScalar, inertiaVector
  , inertia, inertiaVec
    -- * Bounce
  , bouncePhysics
  , bounceScalar, bounceVector
  , bounce, bounceVec
    -- * Elastic
  , elasticOscillation
  , elasticScalar, elasticVector
  , elastic, elasticVec
    -- * Helpers
  , safeAmplitude, safeFrequency, safeDecay
  , safeElasticity, safeGravity, safePeriod
  ) where

import Data.Vector (Vector)
import qualified Data.Vector as V

--------------------------------------------------------------------------------
-- Constants
--------------------------------------------------------------------------------

-- | Default inertia amplitude
defaultAmplitude :: Double
defaultAmplitude = 0.1

-- | Default inertia frequency
defaultFrequency :: Double
defaultFrequency = 2.0

-- | Default inertia decay
defaultDecay :: Double
defaultDecay = 2.0

-- | Default bounce elasticity
defaultElasticity :: Double
defaultElasticity = 0.7

-- | Default bounce gravity
defaultGravity :: Double
defaultGravity = 4000.0

-- | Default elastic period
defaultPeriod :: Double
defaultPeriod = 0.3

-- | Maximum bounces to simulate
maxBounces :: Int
maxBounces = 10

--------------------------------------------------------------------------------
-- Helper Functions
--------------------------------------------------------------------------------

-- | Ensure amplitude is valid
safeAmplitude :: Double -> Double
safeAmplitude amp = if isFinite amp then amp else defaultAmplitude
  where isFinite x = not (isNaN x || isInfinite x)

-- | Ensure frequency is valid
safeFrequency :: Double -> Double
safeFrequency freq = if isFinite freq then freq else defaultFrequency
  where isFinite x = not (isNaN x || isInfinite x)

-- | Ensure decay is valid (minimum 0.001)
safeDecay :: Double -> Double
safeDecay decay = if isFinite decay then max 0.001 decay else defaultDecay
  where isFinite x = not (isNaN x || isInfinite x)

-- | Ensure elasticity is valid (clamped to [0, 1])
safeElasticity :: Double -> Double
safeElasticity e = if isFinite e then min 1 (max 0 e) else defaultElasticity
  where isFinite x = not (isNaN x || isInfinite x)

-- | Ensure gravity is valid (positive)
safeGravity :: Double -> Double
safeGravity g = if isFinite g && g > 0 then g else defaultGravity
  where isFinite x = not (isNaN x || isInfinite x)

-- | Ensure period is valid (positive)
safePeriod :: Double -> Double
safePeriod p = if isFinite p && p > 0 then p else defaultPeriod
  where isFinite x = not (isNaN x || isInfinite x)

--------------------------------------------------------------------------------
-- Inertia Oscillation
--------------------------------------------------------------------------------

-- | Damped sine wave oscillation for inertia effect.
--
-- velocity: Initial velocity component
-- amplitude: Oscillation amplitude multiplier
-- frequency: Oscillation frequency in Hz
-- decay: Exponential decay rate
-- t: Time since keyframe (must be positive)
--
-- Returns: Oscillation offset to add to value
inertiaOscillation :: Double -> Double -> Double -> Double -> Double -> Double
inertiaOscillation velocity amplitude frequency decay t
  | t <= 0 = 0
  | decayFactor == 0 = 0
  | otherwise = velocity * safeAmp * sin phase / decayFactor
  where
    safeAmp = safeAmplitude amplitude
    safeFreq = safeFrequency frequency
    safeDec = safeDecay decay
    phase = safeFreq * t * 2 * pi
    decayFactor = exp (safeDec * t)

-- | Apply inertia to a scalar value
inertiaScalar :: Double -> Double -> Double -> Double -> Double -> Double -> Double
inertiaScalar value velocity amplitude frequency decay t =
  value + inertiaOscillation velocity amplitude frequency decay t

-- | Apply inertia to a vector value
inertiaVector :: Vector Double -> Vector Double -> Double -> Double -> Double -> Double -> Vector Double
inertiaVector values velocities amplitude frequency decay t =
  V.zipWith (\v vel -> v + inertiaOscillation vel amplitude frequency decay t) values velocities

--------------------------------------------------------------------------------
-- Bounce Physics
--------------------------------------------------------------------------------

-- | Calculate bounce position given time and parameters.
--
-- t: Time since keyframe (must be positive)
-- elasticity: Bounce energy retention (0-1)
-- gravity: Gravity acceleration
--
-- Returns: Bounce offset to subtract from value
bouncePhysics :: Double -> Double -> Double -> Double
bouncePhysics t elasticity gravity
  | t <= 0 = 0
  | otherwise = bounceOffset * (1 - safeE)
  where
    safeE = safeElasticity elasticity
    safeG = safeGravity gravity

    -- Calculate which bounce we're in
    findBounce 0 bounceTime height = (bounceTime, height)
    findBounce n bounceTime height
      | bounceTime <= 0 = (bounceTime, height)
      | bounceTime < bounceDuration * 2 = (bounceTime, height)
      | otherwise = findBounce (n - 1) newTime newHeight
      where
        bounceDuration = sqrt (2 * height / safeG)
        newTime = bounceTime - bounceDuration * 2
        newHeight = height * safeE * safeE

    (bounceTime, bounceHeight) = findBounce maxBounces t 1.0

    -- Position within current bounce (parabola)
    bounceDuration = sqrt (2 * bounceHeight / safeG)
    bounceT = if bounceDuration == 0 then 0 else bounceTime / (bounceDuration * 2)
    bounceOffset = bounceHeight * 4 * bounceT * (1 - bounceT)

-- | Apply bounce to a scalar value
bounceScalar :: Double -> Double -> Double -> Double -> Double
bounceScalar value t elasticity gravity =
  value - bouncePhysics t elasticity gravity

-- | Apply bounce to a vector value (same offset for all components)
bounceVector :: Vector Double -> Double -> Double -> Double -> Vector Double
bounceVector values t elasticity gravity =
  let offset = bouncePhysics t elasticity gravity
  in V.map (\v -> v - offset) values

--------------------------------------------------------------------------------
-- Elastic Oscillation
--------------------------------------------------------------------------------

-- | Elastic spring-like oscillation.
--
-- t: Time since keyframe (must be positive)
-- amplitude: Maximum displacement
-- period: Oscillation period in seconds
--
-- Returns: Oscillation offset to add to value
elasticOscillation :: Double -> Double -> Double -> Double
elasticOscillation t amplitude period
  | t <= 0 = 0
  | otherwise = safeAmp * decay * sin phase
  where
    safeAmp = safeAmplitude amplitude
    safePer = safePeriod period
    s = safePer / 4
    -- Exponential decay: 2^(-10*t)
    decay = exp (-10 * t * log 2)
    phase = (t - s) * 2 * pi / safePer

-- | Apply elastic motion to a scalar value
elasticScalar :: Double -> Double -> Double -> Double -> Double
elasticScalar value t amplitude period =
  value + elasticOscillation t amplitude period

-- | Apply elastic motion to a vector value
elasticVector :: Vector Double -> Double -> Double -> Double -> Vector Double
elasticVector values t amplitude period =
  let oscillation = elasticOscillation t amplitude period
  in V.map (+ oscillation) values

--------------------------------------------------------------------------------
-- Main API
--------------------------------------------------------------------------------

-- | Inertia expression for scalar values
inertia :: Double -> Double -> Double -> Double -> Double -> Double -> Double
inertia value velocity t amplitude frequency decay =
  inertiaScalar value velocity amplitude frequency decay t

-- | Inertia expression for vector values
inertiaVec :: Vector Double -> Vector Double -> Double -> Double -> Double -> Double -> Vector Double
inertiaVec values velocities t amplitude frequency decay =
  inertiaVector values velocities amplitude frequency decay t

-- | Bounce expression for scalar values
bounce :: Double -> Double -> Double -> Double -> Double
bounce value t elasticity gravity =
  bounceScalar value t elasticity gravity

-- | Bounce expression for vector values
bounceVec :: Vector Double -> Double -> Double -> Double -> Vector Double
bounceVec values t elasticity gravity =
  bounceVector values t elasticity gravity

-- | Elastic expression for scalar values
elastic :: Double -> Double -> Double -> Double -> Double
elastic value t amplitude period =
  elasticScalar value t amplitude period

-- | Elastic expression for vector values
elasticVec :: Vector Double -> Double -> Double -> Double -> Vector Double
elasticVec values t amplitude period =
  elasticVector values t amplitude period
